timelock/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+//! This contract demonstrates 'timelock' concept and implements a
+//! greatly simplified Claimable Balance (similar to
+//! https://developers.stellar.org/docs/glossary/claimable-balance).
+//! The contract allows to deposit some amount of token and allow another
+//! account(s) claim it before or after provided time point.
+#![no_std]
+#[cfg(feature = "testutils")]
+extern crate std;
+
+use soroban_auth::{
+    check_auth, NonceAuth, {Identifier, Signature},
+};
+use soroban_sdk::{contractimpl, contracttype, BigInt, BytesN, Env, IntoVal, Symbol, Vec};
+
+mod token {
+    soroban_sdk::contractimport!(file = "../soroban_token_contract.wasm");
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum DataKey {
+    Balance,
+    Nonce(Identifier),
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum TimeBoundKind {
+    Before,
+    After,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct TimeBound {
+    pub kind: TimeBoundKind,
+    pub timestamp: u64,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct ClaimableBalance {
+    pub token: BytesN<32>,
+    pub amount: BigInt,
+    pub claimants: Vec<Identifier>,
+    pub time_bound: TimeBound,
+}
+
+pub struct ClaimableBalanceContract;
+
+// The 'timelock' part: check that provided timestamp is before/after
+// the current ledger timestamp.
+fn check_time_bound(env: &Env, time_bound: &TimeBound) -> bool {
+    let ledger_timestamp = env.ledger().timestamp();
+
+    match time_bound.kind {
+        TimeBoundKind::Before => ledger_timestamp <= time_bound.timestamp,
+        TimeBoundKind::After => ledger_timestamp >= time_bound.timestamp,
+    }
+}
+
+// Contract usage pattern (pseudocode):
+// 1. Depositor calls `token.approve(depositor_auth, claimable_balance_contract, 100)`
+//    to allow contract to withdraw the needed amount of token.
+// 2. Depositor calls `deposit(depositor_auth, token_id, 100, claimants, time bound)`. Contract
+//    withdraws the provided token amount and stores it until one of the claimants
+//    claims it.
+// 3. Claimant calls `claim(claimant_auth)` and if time/auth conditons are passed
+//    receives the balance.
+#[contractimpl]
+impl ClaimableBalanceContract {
+    pub fn deposit(
+        env: Env,
+        from: Signature,
+        token: BytesN<32>,
+        amount: BigInt,
+        claimants: Vec<Identifier>,
+        time_bound: TimeBound,
+    ) {
+        if claimants.len() > 10 {
+            panic!("too many claimants");
+        }
+
+        let from_id = from.get_identifier(&env);
+        // Authenticate depositor with nonce of zero, so that this may
+        // be successfully called just once.
+        check_auth(
```

Comment:
what happens if deposit is called twice by different depositors? it seems that the second call will overwrite `DataKey::Balance`?

Reply:
There's a check at the top of `deposit`, a call to `is_initialized`, which could probably have a better name and be named something like `has_deposit`.

Reply:
@leighmcculloch I see `is_initialized()` is called from the `claim()` but I don't think it's called from `deposit()`

Reply:
Ah, wait, @dmkozh I think the is_initialized got removed from teh wrong function. It should have been removed from claim, not from deposit.

Reply:
btw, I think it's possible to have an alternative design where multiple claimable balances are stored within 1 contract rather than having a separate contract per claimable balance. 

This could be implemented by adding an extra parameter to the `deposit()` and `claim()` functions which represents the claimable balance's id (or maybe the id can be derived by hashing the claimable balance + a salt). The balance would be mapped to the id in storage.  I think the deposit flow would be cheaper in this design because the depositor would not need to deploy a new smart contract.

Reply:
> I think it's possible to have an alternative design where multiple claimable balances are stored within 1 contract rather than having a separate contract per claimable balance.

Indeed, I've considered that option as well. However, it goes beyond the point of the example (which is about interacting with the ledger state in the first place) and is more complex. FWIW, the same principle applies to most of the examples (e.g. one could have a contract that manages several liquidity pools or offers). This has been discussed on discord as well:  https://discord.com/channels/897514728459468821/1013959254245388430

Reply:
One important con of a 'single contract-multiple entities' approach is that we need to either block on a single autoincrement entry for producing the ids or need to introduce some good way of producing the unique ids deterministically (hash is not enough as one may have multiple entries with the same content).